### PR TITLE
Bugfix/zenodo connector rate limits

### PR DESCRIPTION
**Pull request by jsmatias**
Bugfixes:
1. The connector was unable to fetch all records and process them before the resumption token expiration. (422 error from zenodo). This is related to the issue #267 
     Solution: I separated fetching all records from processing them, and added a timer to stop fetching before the given expiration. 
2. The previous bug resulted in skipping several records.
     Solution: I pass the timestamp of the last processed record to the state. This way the next iteration will start from the last record, instead of `to_excl_current`, which is defined as next cycle (previous `from_incl + time_per_loop`)
3. Zenodo connector was exceeding the request rate limits of the Zenodo platform. This is related to the issue #267 
      Solution: I used decorators on the fetching methods to define the limits and sleep and retry to avoid raising the `429 error`. 
4. Many `KeyError`s due to missing fields were not properly addressed. This were all passed to the log files making it less readable to debug and quickly overloads the log files. 
     Solution: I used the dictionary `.get()` method to return a `None` or default value to be evaluated subsequently. 

Changes:
1. I included a few unit tests to test these bugfixes
2. Changed the initial date from which the connector starts fetching data 2023 -> 2020

### CODE REVIEW

File: src/connectors/zenodo/zenodo_dataset_connector.py [hunk 1]
```diff
@@ -54,25 +61,44 @@ def _error_msg_bad_format(field) -> str:
...
             return RecordError(identifier=identifier, error=error_fmt("creator"))
 
         creators = []
         pydantic_class_contact = resource_create(Contact)
         for name in creator_names:
             creators.append(pydantic_class_contact(name=name))
 
-        if isinstance(record["titles"]["title"], str):
-            title = record["titles"]["title"]
+        titles_raw = record.get("titles", {}).get("title")
```
Comment (andrejridzik) [inline src/connectors/zenodo/zenodo_dataset_connector.py:99]: Aren't we expecting a list of titles here? If not, maybe we can add a comment that a single title is expected always

File: src/connectors/zenodo/zenodo_dataset_connector.py [hunk 2]
```diff
@@ -184,37 +225,130 @@ def _resource_type(record) -> str | None:
...
                 try:
                     xml_string = record.raw
                     xml_dict = xmltodict.parse(xml_string)
                     id_ = xml_dict["record"]["header"]["identifier"]
                     if id_.startswith("oai:"):
                         id_ = id_.replace("oai:", "")
-                    datetime_ = dateutil.parser.parse(xml_dict["record"]["header"]["datestamp"])
-                    resource = xml_dict["record"]["metadata"]["oai_datacite"]["payload"]["resource"]
-                    yield datetime_, self._dataset_from_record(id_, resource)
+                    # datetime_ = dateutil.parser.parse(xml_dict["record"]["header"]["datestamp"])
```
Comment (andrejridzik) [inline src/connectors/zenodo/zenodo_dataset_connector.py:339]: I assume this line can be removed now

Comment (jsmatias) [reply src/connectors/zenodo/zenodo_dataset_connector.py:339]: Well noticed! Thanks.

File: src/tests/connectors/zenodo/test_get_datasets_zenodo.py [hunk 3]
```diff
@@ -58,25 +86,166 @@ def test_fetch_happy_path():
...
+        state = {}
+        from_incl = datetime.datetime(2023, 5, 23, 8, 0, 0)
+        to_excl = datetime.datetime(2023, 5, 23, 9, 0, 0)
+        resources = list(connector.run(state, from_incl=from_incl, to_excl=to_excl))
+        datasets = [r for r in resources if not isinstance(r, RecordError)]
+        errors = [r for r in resources if isinstance(r, RecordError)]
+
+        assert {error.error for error in errors} == {"Wrong type"}
+        assert len(datasets) == 6
+        assert len(errors) == 45
```
[the remark this answers is not in the document]
Comment (jsmatias) [reply src/tests/connectors/zenodo/test_get_datasets_zenodo.py:116]: No, actually the description is wrong. Once the token is about to expires, it is expected that the connector stops fetching more lists (or pages). Then it processes all the fetched records and stores the last one's `datestamp` to resume the requests during next iteration. 
So 20 + 5 ('wrong type' + 'datasets') are the records from the first list. Once the connector finishes processing these ones, it should resume requesting more records from the next list with a new token. At the end of the process it fetches and processes all records (6 + 45) from both lists without skipping any due to token expiration.

File: src/connectors/zenodo/zenodo_dataset_connector.py (hunk 1 above)
Comment (jsmatias) [reply src/connectors/zenodo/zenodo_dataset_connector.py:99]: Not really, each dataset should have a single title. 
I added the following comment:
# Checking only the str type, as each dataset is expected to have a single title.

**Review by andrejridzik — APPROVED** [review]

**Pull request merged**